### PR TITLE
Draft: [ME] geographical extent

### PR DESCRIPTION
**Pull request by Angi-Kinas**
### Description

This PR introduces [...]

### Architectural changes

The following library now depends on [...]

### Screenshots

[...]

### Quality Assurance Checklist

- [ ] Commit history is devoid of any _merge commits_ and readable to facilitate reviews
- [ ] If **new logic** ⚙️ is introduced: unit tests were added
- [ ] If **new user stories** 🤏 are introduced: E2E tests were added
- [ ] If **new UI components** 🕹️ are introduced: corresponding stories in Storybook were created
- [ ] If **breaking changes** 🪚 are introduced: add the `breaking change` label
- [ ] If **bugs** 🐞 are fixed: add the `backport <release branch>` label
- [ ] The [documentation website](docs) 📚 has received the love it deserves

---

**This work is sponsored by [Organization ABC](xx)**.

### CODE REVIEW

File: libs/common/domain/src/lib/model/record/metadata.model.ts [hunk 1]
```diff
@@ -159,7 +165,8 @@ export interface GraphicOverview {
 }
 
 export interface DatasetSpatialExtent {
-  geometry: Geometry
+  geometries?: Geometry[]
```
Comment (Angi-Kinas) [inline libs/common/domain/src/lib/model/record/metadata.model.ts:168]: This will be part of https://github.com/geonetwork/geonetwork-ui/pull/944

File: libs/api/repository/src/lib/gn4/platform/gn4-platform.mapper.ts [hunk 2]
```diff
@@ -90,6 +90,12 @@ export class Gn4PlatformMapper {
           url: new URL(matchedThesaurus?.url),
           type: matchedThesaurus?.dname as KeywordType,
         },
+        coords: {
+          coordEast: keyword.coordEast,
```
Comment (Angi-Kinas) [inline libs/api/repository/src/lib/gn4/platform/gn4-platform.mapper.ts:94]: The coords are saved as strings. Should the parsing to a number already happen here?
And maybe already return a bbox instead?

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import { of } from 'rxjs'
+import { FormControl } from '@angular/forms'
+import { PlatformServiceInterface } from '@geonetwork-ui/common/domain/platform.service.interface'
+
+class PlatformServiceInterfaceMock {
+  searchKeywords = jest.fn(() =>
+    of([{ label: 'Address', thesaurus: { id: '1' } }])
+  )
+}
+describe('FormFieldKeywordsComponent', () => {
```
Comment (Angi-Kinas) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.spec.ts:19]: To do: Write tests for this component

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+} from '@geonetwork-ui/ui/inputs'
+import { UiWidgetsModule } from '@geonetwork-ui/ui/widgets'
+import { map } from 'rxjs'
+import { Keyword } from '@geonetwork-ui/common/domain/model/record'
+import { KeywordType } from '@geonetwork-ui/common/domain/model/thesaurus'
+
+type AutocompleteItem = { title: string; value: Keyword }
+
+@Component({
+  selector: 'gn-ui-form-field-keywords-generic',
```
Comment (Angi-Kinas) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.ts:24]: Change name to `gn-ui-form-field-generic-keywords` as the name of this component. 
Are there better ideas for naming?

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  standalone: true,
+  imports: [
+    DropdownSelectorComponent,
+    UiInputsModule,
+    CommonModule,
+    UiWidgetsModule,
+    AutocompleteComponent,
+  ],
+})
+export class GenericFormFieldKeywordsComponent {
```
Comment (Angi-Kinas) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.ts:37]: Use this component also for the keywords field (other than places)

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { FormFieldMapContainerComponent } from './form-field-map-container.component';
+
+describe('FormFieldMapContainerComponent', () => {
```
Comment (Angi-Kinas) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.spec.ts:5]: To do: write tests for this component

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.spec.ts [hunk 7]
```diff
@@ -1,18 +1,93 @@
...
+}
+class MapFacadeMock {
+  addLayer = jest.fn()
+}
+class PlatformServiceInterfaceMock {
+  searchKeywords = jest.fn(() =>
+    of([{ label: 'Africa', thesaurus: { id: '1' } }])
+  )
+}
 describe('FormFieldSpatialExtentComponent', () => {
```
Comment (Angi-Kinas) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.spec.ts:43]: To do: write tests for this component

File: libs/feature/editor/src/lib/components/record-form/record-form.component.ts [hunk 8]
```diff
@@ -2,20 +2,99 @@ import { CommonModule } from '@angular/common'
...
   EditorFieldWithValue,
   EditorSectionWithValues,
 } from '../../+state/editor.models'
+import {
+  DatasetSpatialExtent,
+  Keyword,
+} from '@geonetwork-ui/common/domain/model/record'
+import { KeywordType } from '@geonetwork-ui/common/domain/model/thesaurus'
 
+const DUMMY_DATA_PLACE_KEYWORDS = [
```
Comment (Angi-Kinas) [inline libs/feature/editor/src/lib/components/record-form/record-form.component.ts:20]: Remove dummy data once it has been tested.

File: libs/feature/editor/src/lib/components/record-form/record-form.component.ts [hunk 9]
```diff
@@ -34,4 +113,31 @@ export class RecordFormComponent {
...
+
+  extractSpatialExtentsFields(fieldsWithValues: EditorFieldWithValue[]) {
+    const placeKeywordsField = fieldsWithValues.find(
+      (field) => field.config.id === 'placeKeywords'
+    )?.value as Keyword[]
+    const spatialExtentsField = fieldsWithValues.find(
+      (field) => field.config.model === 'spatialExtents'
+    )?.value as DatasetSpatialExtent[]
+    if (spatialExtentsField && placeKeywordsField) {
+      return {
```
Comment (Angi-Kinas) [inline libs/feature/editor/src/lib/components/record-form/record-form.component.ts:135]:
```suggestion
      return {
        placeKeywordsField,
        spatialExtentsField
      }
```

Comment (Angi-Kinas) [reply libs/feature/editor/src/lib/components/record-form/record-form.component.ts:135]: Edit once the dummy data is gone

File: libs/feature/map/src/lib/utils/map-utils.service.ts [hunk 10]
```diff
@@ -223,7 +223,7 @@ export class MapUtilsService {
     // transform an array of geojson geometries into a bbox
     const totalExtent = record.spatialExtents.reduce(
       (prev, curr) => {
-        const geom = GEOJSON.readGeometry(curr.geometry)
+        const geom = GEOJSON.readGeometry(curr.geometries[0])
```
Comment (Angi-Kinas) [inline libs/feature/map/src/lib/utils/map-utils.service.ts:226]: Check if this is correct. Depends on https://github.com/geonetwork/geonetwork-ui/pull/944

**Review by Angi-Kinas — COMMENTED** [review]
Thanks @LHBruneton-C2C  for your help on this!

Please feel free to review and leave comments.
I already left some comments to not forget what still needs to be done.
Another ting that I was unsure about, is where to place the components that have been created. Namely "GenericFormFieldKeywordsComponent" and "FormFieldMapContainerComponent" and about their naming.

Thanks in advance!

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.ts (hunk 4 above)
Comment (LHBruneton-C2C) [reply libs/feature/editor/src/lib/components/record-form/form-field/form-field-keywords-generic/form-field-keywords-generic.component.ts:24]: I'm not really sure what this component is responsible of, as the keywords have already been dealt with in an another already merged PR. If this is the common part between the two features, this should be removed from the other form-field (didn't see any dropped code in this PR). And you could place it in feature/editor/src/lib/components, and the drop the "form-field" part of the name.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.spec.ts (hunk 6 above)
Comment (LHBruneton-C2C) [reply libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.spec.ts:5]: Maybe this could be moved to feature/editor/src/lib too, to split things a bit more.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        if (
+          this.keywordsWithSpatialExtents[key].spatialExtents?.geometries
+            ?.length >= 0
+        ) {
+          this.keywordsWithSpatialExtents[
+            key
+          ].spatialExtents.geometries.forEach((geometry) =>
+            featureCollection.features.push({
+              type: 'Feature',
+              properties: { description: key },
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts:111]: This "properties" field is not really needed for the map preview. It would be, if at some point we were allowing the user to click on a polygon and have a popup showing the description, for example.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        [bbox[0], bbox[3]],
+        [bbox[2], bbox[3]],
+      ],
+    ])
+
+    const geoJSONGeom = new GeoJSON().writeGeometryObject(geometry)
+    return geoJSONGeom
+  }
+
+  addToMap(
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts:149]: Where is this used? The name doesn't really match the content of the function, and I couldn't find any reference.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+  styleUrls: ['./form-field-map-container.component.css'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class FormFieldMapContainerComponent implements OnChanges {
+  @Input() keywordsWithSpatialExtents: {
+    [key: string]: {
+      placeKeyword?: Keyword
+      spatialExtents?: DatasetSpatialExtent
+    }
+  }[]
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts:45]: Not critical, but this component doesn't need to know anything about keywords, so it would be better if you only gave it an array of spatial extents as input. This could be done by reducing/flattening your keywordsWithSpatialExtents from the parent component before sending it to the child component.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.ts [hunk 14]
```diff
@@ -1,10 +1,146 @@
...
+
+  ngOnInit(): void {
+    this.linkPlaceKeywordsToSpatialExtents(
+      this.placeKeywords,
+      this.spatialExtents
+    )
+    this.linkSpatialExtentsToPlaceKeywords(
+      this.spatialExtents,
+      this.placeKeywords
+    )
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.ts:66]: Good call, splitting this in subfunctions :+1:

File: libs/api/repository/src/lib/gn4/platform/gn4-platform.mapper.ts (hunk 2 above)
Comment (LHBruneton-C2C) [reply libs/api/repository/src/lib/gn4/platform/gn4-platform.mapper.ts:94]: The way you did it is fine, I'm not sure we will need those coords as string anywhere else yet.

As for the order of the coords, were did you get it come from? In an object it doesn't really matter, the properties are not ordered, but later on I see you have to switch to arrays, and the order is really hard to follow.

In GIS, the most frequent order is [minx, miny, maxx, maxy] which is the equivalent of [westBound, southBound, eastBound, northBound].

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+        type: MapContextLayerTypeEnum.GEOJSON,
+        data: featureCollection,
+        title: 'Spatial extents',
+      })
+    }
+  }
+  bboxCoordsToGeometry(bbox: [number, number, number, number]): Geometry {
+    const geometry = new Polygon([
+      [
+        [bbox[2], bbox[3]],
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts:137]: This is original, why not start with 0 - 1?

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.ts [hunk 16]
```diff
@@ -1,10 +1,146 @@
...
+    FormFieldMapContainerComponent,
+  ],
 })
-export class FormFieldSpatialExtentComponent {}
+export class FormFieldSpatialExtentComponent implements OnInit {
+  @Input() placeKeywords: Keyword[]
+  @Input() spatialExtents: DatasetSpatialExtent[]
+
+  @Output() placeKeywordChange: EventEmitter<Keyword[]> = new EventEmitter()
+  @Output() spatialExtentsChange: EventEmitter<DatasetSpatialExtent[]> =
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.ts:48]: Do you intend to output in the next PR? This could be part of this PR already.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.ts [hunk 17]
```diff
@@ -1,10 +1,146 @@
...
+        }
+      }
+    })
+
+    const missingPlaces = Object.keys(this.keywordsLinkedToExtents).filter(
+      (uri) => !placeKeywords.some(({ key: id }) => uri === id)
+    )
+
+    missingPlaces.forEach((missingPlace) => {
+      this.placeKeywords.push(
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.ts:136]: I don't think we fix inconsistencies in a record if there are more spatialExtents than keywords. This link between place keywords and spatial extents is our custom way of doing things, not any standard.

We will push new place keywords and new spatial extents as the user adds them from our interface, but we will not push from what we've read from the record.

**Review by LHBruneton-C2C — COMMENTED** [review]
Nice work already! Not tested yet, will try tomorrow.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts (hunk 11 above)
Comment (Angi-Kinas) [reply libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts:111]: Typescript complains when I remove the properties. I could leave the object empty though.
```suggestion
              properties: {},
```

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts (hunk 12 above)
Comment (Angi-Kinas) [reply libs/feature/editor/src/lib/components/record-form/form-field/form-field-map-container/form-field-map-container.component.ts:149]: Not used anymore

**Comment by jahow** [comment]
@LHBruneton-C2C @Angi-Kinas I finally managed to rework this, I have rewritten the commit history to make it more readable and hopefully easier to review.

Things I added to the original work:
* make it so that the `form-field-spatial-extent` component directly reads from and writes to the editor facade; this makes it not necessary to have an elaborate logic in the `record-form` component because this field works with two properties (keywords and extents) instead of one
* fixed the missing translation bug (was due to an erroneous import of `TranslateModule.forRoot()` in ui-map which caused another translate store to be instanced)
* reworked the `form-field-spatial-extent` to make it more straightforward: the link between extents and keywords is now done directly on the keyword (by storing it in a `_linkedExtent` property); this works because the user only interacts with keywords so then when one is added/removed we can add and remove extents accordingly
* wrote missing tests
* added a `searchKeywordsByThesaurus` in platform service; this is useful because keywords coming from the XML record do not have a bbox or key attached; this information has to be read from the thesaurus API
* added a missing `ol.css` file

This is ready for review, I hope I didn't miss anything.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.spec.ts [hunk 18]
```diff
@@ -1,16 +1,148 @@
...
+          record$: new BehaviorSubject(SAMPLE_RECORD),
+          updateRecordField: jest.fn(),
+        }),
+      ],
+    })
+      .overrideComponent(FormFieldSpatialExtentComponent, {
+        set: {
+          changeDetection: ChangeDetectionStrategy.Default,
+        },
+      })
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field-spatial-extent/form-field-spatial-extent.component.spec.ts:142]: I don't think this is necessary anymore with the MockBuilder.

File: libs/feature/editor/src/lib/components/record-form/form-field/form-field.component.spec.ts [hunk 19]
```diff
@@ -14,32 +13,25 @@ import { FormFieldOverviewsComponent } from './form-field-overviews/form-field-o
...
+  })
+
   beforeEach(async () => {
     await TestBed.configureTestingModule({
-      imports: [
-        FormFieldComponent,
-        TranslateModule.forRoot(),
-        HttpClientTestingModule,
-      ],
-      schemas: [NO_ERRORS_SCHEMA],
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/components/record-form/form-field/form-field.component.spec.ts:37]: Can't wait to see more of those disappear!

File: libs/feature/editor/src/lib/models/editor-config.model.ts [hunk 20]
```diff
@@ -17,6 +17,7 @@ export interface FormFieldConfig {
 }
 
 export interface EditorField {
+  id?: string
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/lib/models/editor-config.model.ts:20]: This is not needed anymore, since you used the facade to retrieve multiple fields.

File: libs/feature/editor/src/test-setup.ts [hunk 21]
```diff
@@ -6,10 +6,25 @@ import {
...
   platformBrowserDynamicTesting(),
   { teardown: { destroyAfterEach: false } }
 )
+
+// ng-mocks global configuration
+ngMocks.autoSpy('jest')
+
+ngMocks.globalKeep(CommonModule, true)
+ngMocks.globalKeep(BrowserModule, true)
+ngMocks.globalKeep(RouterModule, true)
```
Comment (LHBruneton-C2C) [inline libs/feature/editor/src/test-setup.ts:28]: Not sure the Router should be kept in libs.

**Review by LHBruneton-C2C — APPROVED** [review]
Looking great! Thanks @Angi-Kinas and @jahow for this huge feature.
Tested draft and publishing, working as it should.

**Review by Angi-Kinas — COMMENTED** [review]
Thank you very much @jahow for making this work! It was a really big and complicated feature.

I tested the feature a bit and I think we still need to filter out the place keywords from the "other" keywords that we display on the 3rd page of the editor.
Otherwise we get issues like this:
image
And when we delete the keyword "Africa" and go back to the first page, we see:
image
WDYT?

**Comment by jahow** [comment]
@Angi-Kinas yes you're right... could this be done in another PR? because I honestly don't feel like adding anything else to this one :scream:

**Comment by Angi-Kinas** [comment]
@jahow yes you are right, I totally understand! Let's do it in another PR.
